// @tria-sdk/web/services/auth.ts
import BackendService from "./index";
import { errorHandler } from "../utils/helper";

class AuthService extends BackendService {
    constructor(apiUrl: string) {
        super(apiUrl);
    }

    async initiateOTP({ input }: { input: string }) {
        try {
            const { data } = await this.triaAxios.post("/auth/otp/initiate", {
                input
            });
            return data;
        } catch (err) {
            return errorHandler(err);
        }
    }

    async verifyOTP({ otp, session, input }: { input: string, session: string, otp: string }) {
        try {
            const { data } = await this.triaAxios.post("/auth/otp/verify", {
                otp,
                session,
                input
            });
            return data;
        } catch (err) {
            return errorHandler(err);
        }
    }

    async getSession() {
        try {
            const { data } = await this.triaAxios.get("/auth/issue-token");
            return data;
        } catch (err) {
            return errorHandler(err);
        }
    }

    async getUserSession() {
        try {
            const { data } = await this.triaAxios.get("/auth/get-session");
            return data;
        } catch (err) {
            return errorHandler(err);
        }
    }

    async getKey() {
        try {
            const { data } = await this.triaAxios.get("/auth/get-key");
            return data;
        } catch (err) {
            return errorHandler(err);
        }
    }
}

export default AuthService;
