// @tria-sdk/web/services/nonevm.ts
import { Send, Swap } from "../types";
import BackendService from "./index";
import { errorHandler } from "../utils/helper";
class nonEVMService extends BackendService {
    constructor(apiUrl: string) {
        super(apiUrl);
    }

    async send(shard1: string, chainName: string, payload: Send) {
        try {
            const { data } = await this.triaAxios.post("/nonevm/send", {
                shard1: shard1,
                chainName: chainName,
                payload: payload
            })
            return data;
        } catch (err: unknown) {
            return errorHandler(err);
        }
    }

    async swap(shard1: string, chainName: string, payload: Swap) {
        try {
            const { data } = await this.triaAxios.post("/nonevm/swap", {
                shard1: shard1,
                chainName: chainName,
                payload: payload
            })
            return data;
        } catch (err: unknown) {
            return errorHandler(err);
        }
    }
}

export default nonEVMService;
