// @decloudlabs/stk-v2/lib/ABI/AppDeployment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AppDeployment = [
    {
        inputs: [
            {
                internalType: "uint256[]",
                name: "nftIDList",
                type: "uint256[]",
            },
            {
                internalType: "bytes32[]",
                name: "appName",
                type: "bytes32[]",
            },
            {
                internalType: "bytes[2][][]",
                name: "path",
                type: "bytes[2][][]",
            },
            {
                internalType: "uint256[][]",
                name: "subnetList",
                type: "uint256[][]",
            },
            {
                internalType: "uint8[][][]",
                name: "multiplier",
                type: "uint8[][][]",
            },
            {
                internalType: "uint16[][]",
                name: "resourceCount",
                type: "uint16[][]",
            },
            {
                internalType: "uint16[][]",
                name: "resourceType",
                type: "uint16[][]",
            },
            {
                internalType: "bool[]",
                name: "cidLock",
                type: "bool[]",
            },
        ],
        name: "cloneNFTBatch",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "balanceToAdd",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "bytes32",
                name: "appName",
                type: "bytes32",
            },
            {
                internalType: "bytes[2]",
                name: "path",
                type: "bytes[2]",
            },
            {
                internalType: "uint256[]",
                name: "subnetList",
                type: "uint256[]",
            },
            {
                internalType: "uint8[][]",
                name: "multiplier",
                type: "uint8[][]",
            },
            {
                internalType: "uint16[]",
                name: "resourceCount",
                type: "uint16[]",
            },
            {
                internalType: "uint16[]",
                name: "resourceType",
                type: "uint16[]",
            },
            { internalType: "bool", name: "cidLock", type: "bool" },
        ],
        name: "createApp",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "balanceToAdd",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "bytes32[]",
                name: "appName",
                type: "bytes32[]",
            },
            {
                internalType: "bytes[2][]",
                name: "path",
                type: "bytes[2][]",
            },
            {
                internalType: "uint256[][]",
                name: "subnetList",
                type: "uint256[][]",
            },
            {
                internalType: "uint8[][][]",
                name: "multiplier",
                type: "uint8[][][]",
            },
            {
                internalType: "uint16[][]",
                name: "resourceCount",
                type: "uint16[][]",
            },
            {
                internalType: "uint16[][]",
                name: "resourceType",
                type: "uint16[][]",
            },
            {
                internalType: "bool[]",
                name: "cidLock",
                type: "bool[]",
            },
        ],
        name: "createAppBatch",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "appID",
                type: "uint256",
            },
        ],
        name: "deleteApp",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "getActiveSubnetsOfNFT",
        outputs: [{ internalType: "uint256[]", name: "", type: "uint256[]" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "appID",
                type: "uint256",
            },
        ],
        name: "getApp",
        outputs: [
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "appID",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256[]",
                        name: "subnetList",
                        type: "uint256[]",
                    },
                    {
                        internalType: "uint8[][]",
                        name: "currentReplica",
                        type: "uint8[][]",
                    },
                    {
                        components: [
                            {
                                internalType: "bytes32",
                                name: "appName",
                                type: "bytes32",
                            },
                            {
                                internalType: "uint256",
                                name: "timestamp",
                                type: "uint256",
                            },
                            {
                                internalType: "bytes[2]",
                                name: "path",
                                type: "bytes[2]",
                            },
                            {
                                internalType: "uint16[]",
                                name: "resourceCount",
                                type: "uint16[]",
                            },
                            {
                                internalType: "uint16[]",
                                name: "resourceType",
                                type: "uint16[]",
                            },
                            {
                                internalType: "bool",
                                name: "cidLock",
                                type: "bool",
                            },
                            {
                                internalType: "bool",
                                name: "active",
                                type: "bool",
                            },
                        ],
                        internalType: "struct AppDeploymentData.App",
                        name: "app",
                        type: "tuple",
                    },
                ],
                internalType: "struct AppDeploymentData.FullAppData",
                name: "",
                type: "tuple",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "getAppList",
        outputs: [
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "appID",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256[]",
                        name: "subnetList",
                        type: "uint256[]",
                    },
                    {
                        internalType: "uint8[][]",
                        name: "currentReplica",
                        type: "uint8[][]",
                    },
                    {
                        components: [
                            {
                                internalType: "bytes32",
                                name: "appName",
                                type: "bytes32",
                            },
                            {
                                internalType: "uint256",
                                name: "timestamp",
                                type: "uint256",
                            },
                            {
                                internalType: "bytes[2]",
                                name: "path",
                                type: "bytes[2]",
                            },
                            {
                                internalType: "uint16[]",
                                name: "resourceCount",
                                type: "uint16[]",
                            },
                            {
                                internalType: "uint16[]",
                                name: "resourceType",
                                type: "uint16[]",
                            },
                            {
                                internalType: "bool",
                                name: "cidLock",
                                type: "bool",
                            },
                            {
                                internalType: "bool",
                                name: "active",
                                type: "bool",
                            },
                        ],
                        internalType: "struct AppDeploymentData.App",
                        name: "app",
                        type: "tuple",
                    },
                ],
                internalType: "struct AppDeploymentData.FullAppData[]",
                name: "",
                type: "tuple[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
        ],
        name: "getComputesOfSubnet",
        outputs: [
            { internalType: "uint16[]", name: "", type: "uint16[]" },
            { internalType: "uint32[]", name: "", type: "uint32[]" },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "appID",
                type: "uint256",
            },
        ],
        name: "getCurrentReplica",
        outputs: [
            {
                internalType: "uint256[]",
                name: "subnetList",
                type: "uint256[]",
            },
            {
                internalType: "uint8[][]",
                name: "currentReplicaList",
                type: "uint8[][]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "getNFTSubnetList",
        outputs: [{ internalType: "uint256[]", name: "", type: "uint256[]" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "appID",
                type: "uint256",
            },
        ],
        name: "getSubnetsOfApp",
        outputs: [{ internalType: "uint256[]", name: "", type: "uint256[]" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "pause",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "paused",
        outputs: [{ internalType: "bool", name: "status", type: "bool" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "appID",
                type: "uint256",
            },
        ],
        name: "setCIDLock",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "maxResourceLength",
                type: "uint256",
            },
        ],
        name: "setMaxResourceLength",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "maxResourceTypePerSubnet",
                type: "uint256",
            },
        ],
        name: "setMaxResourceTypePerSubnet",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "setSubnetLock",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "balanceToAdd",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "address[]",
                name: "rlsAddresses",
                type: "address[]",
            },
            {
                internalType: "uint256[]",
                name: "licenseFactor",
                type: "uint256[]",
            },
        ],
        name: "subscribe",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "balanceToAdd",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "address[]",
                name: "rlsAddresses",
                type: "address[]",
            },
            {
                internalType: "uint256[]",
                name: "licenseFactor",
                type: "uint256[]",
            },
            {
                internalType: "bytes32",
                name: "appName",
                type: "bytes32",
            },
            {
                internalType: "bytes[2]",
                name: "path",
                type: "bytes[2]",
            },
            {
                internalType: "uint256[]",
                name: "subnetList",
                type: "uint256[]",
            },
            {
                internalType: "uint8[][]",
                name: "multiplier",
                type: "uint8[][]",
            },
            {
                internalType: "uint16[]",
                name: "resourceCount",
                type: "uint16[]",
            },
            {
                internalType: "uint16[]",
                name: "resourceType",
                type: "uint16[]",
            },
            { internalType: "bool", name: "cidLock", type: "bool" },
        ],
        name: "subscribeAndCreateApp",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "unpause",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "balanceToAdd",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "appID",
                type: "uint256",
            },
            {
                internalType: "bytes[2]",
                name: "path",
                type: "bytes[2]",
            },
            {
                internalType: "uint256[]",
                name: "subnetList",
                type: "uint256[]",
            },
            {
                internalType: "uint8[][]",
                name: "multiplier",
                type: "uint8[][]",
            },
            {
                internalType: "uint16[]",
                name: "resourceCount",
                type: "uint16[]",
            },
            {
                internalType: "uint16[]",
                name: "resourceType",
                type: "uint16[]",
            },
        ],
        name: "updateApp",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "balanceToAdd",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "uint256[]",
                name: "appID",
                type: "uint256[]",
            },
            {
                internalType: "bytes[2][]",
                name: "path",
                type: "bytes[2][]",
            },
            {
                internalType: "uint256[][]",
                name: "subnetList",
                type: "uint256[][]",
            },
            {
                internalType: "uint8[][][]",
                name: "multiplier",
                type: "uint8[][][]",
            },
            {
                internalType: "uint16[][]",
                name: "resourceCount",
                type: "uint16[][]",
            },
            {
                internalType: "uint16[][]",
                name: "resourceType",
                type: "uint16[][]",
            },
        ],
        name: "updateAppBatch",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "appID",
                type: "uint256",
            },
            {
                internalType: "bytes[2]",
                name: "path",
                type: "bytes[2]",
            },
        ],
        name: "updateCID",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "uint256[]",
                name: "appIDList",
                type: "uint256[]",
            },
            {
                internalType: "uint256[][]",
                name: "subnetList",
                type: "uint256[][]",
            },
            {
                internalType: "uint8[][][]",
                name: "multiplierList",
                type: "uint8[][][]",
            },
            {
                internalType: "uint16[][]",
                name: "resourceList",
                type: "uint16[][]",
            },
            {
                internalType: "uint16[][]",
                name: "resourceType",
                type: "uint16[][]",
            },
        ],
        name: "updateResource",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "contract ISubscription",
                name: "_Subscription",
                type: "address",
            },
            {
                internalType: "contract IApplicationNFT",
                name: "_AppNFT",
                type: "address",
            },
            {
                internalType: "contract ISubscriptionBalance",
                name: "_SubscriptionBalance",
                type: "address",
            },
            {
                internalType: "contract ISubnetCluster",
                name: "_SubnetCluster",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "_MAX_RESOURCE_LEN",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "_MAX_RESOURCE_TYPE_PER_SUBNET",
                type: "uint256",
            },
            {
                internalType: "contract AppDeploymentInternal",
                name: "_AppDepInternal",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "constructor",
    },
    { inputs: [], name: "AppCountExceededError", type: "error" },
    { inputs: [], name: "AppDoesntExistError", type: "error" },
    { inputs: [], name: "AppNameExistError", type: "error" },
    {
        inputs: [],
        name: "DiamondWritable__InvalidInitializationParameters",
        type: "error",
    },
    {
        inputs: [],
        name: "DiamondWritable__RemoveTargetNotZeroAddress",
        type: "error",
    },
    {
        inputs: [],
        name: "DiamondWritable__ReplaceTargetIsIdentical",
        type: "error",
    },
    {
        inputs: [],
        name: "DiamondWritable__SelectorAlreadyAdded",
        type: "error",
    },
    {
        inputs: [],
        name: "DiamondWritable__SelectorIsImmutable",
        type: "error",
    },
    {
        inputs: [],
        name: "DiamondWritable__SelectorNotFound",
        type: "error",
    },
    {
        inputs: [],
        name: "DiamondWritable__SelectorNotSpecified",
        type: "error",
    },
    {
        inputs: [],
        name: "DiamondWritable__TargetHasNoCode",
        type: "error",
    },
    { inputs: [], name: "DuplicateSubnetError", type: "error" },
    {
        inputs: [],
        name: "ERC165Base__InvalidInterfaceId",
        type: "error",
    },
    {
        inputs: [],
        name: "InvalidMultiplierLengthError",
        type: "error",
    },
    { inputs: [], name: "InvalidResourceLengthError", type: "error" },
    { inputs: [], name: "InvalidSubnetError", type: "error" },
    { inputs: [], name: "LockedSubnetListError", type: "error" },
    {
        inputs: [],
        name: "MaxResourceTypeReachedError",
        type: "error",
    },
    { inputs: [], name: "MaxSubnetReachedError", type: "error" },
    { inputs: [], name: "NoAppsExistError", type: "error" },
    { inputs: [], name: "NoCallPermissionError", type: "error" },
    { inputs: [], name: "Ownable__NotOwner", type: "error" },
    {
        inputs: [],
        name: "Ownable__NotTransitiveOwner",
        type: "error",
    },
    { inputs: [], name: "PathLockedError", type: "error" },
    {
        inputs: [],
        name: "Proxy__ImplementationIsNotContract",
        type: "error",
    },
    {
        inputs: [],
        name: "ResourceTypeAndCountMismatchError",
        type: "error",
    },
    {
        inputs: [],
        name: "SafeOwnable__NotNomineeOwner",
        type: "error",
    },
    { inputs: [], name: "SubscriptionNotDoneError", type: "error" },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256[]",
                name: "nftIDList",
                type: "uint256[]",
            },
            {
                indexed: false,
                internalType: "bytes32[]",
                name: "appName",
                type: "bytes32[]",
            },
            {
                indexed: false,
                internalType: "bytes[2][][]",
                name: "path",
                type: "bytes[2][][]",
            },
            {
                indexed: false,
                internalType: "uint256[][]",
                name: "subnetList",
                type: "uint256[][]",
            },
            {
                indexed: false,
                internalType: "uint8[][][]",
                name: "multiplier",
                type: "uint8[][][]",
            },
            {
                indexed: false,
                internalType: "uint16[][]",
                name: "resourceCount",
                type: "uint16[][]",
            },
            {
                indexed: false,
                internalType: "uint16[][]",
                name: "resourceType",
                type: "uint16[][]",
            },
            {
                indexed: false,
                internalType: "bool[]",
                name: "cidLock",
                type: "bool[]",
            },
        ],
        name: "AppNFTCloned",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "appID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "bytes32",
                name: "appName",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "bytes[2]",
                name: "path",
                type: "bytes[2]",
            },
            {
                indexed: false,
                internalType: "uint256[]",
                name: "subnetList",
                type: "uint256[]",
            },
            {
                indexed: false,
                internalType: "uint8[][]",
                name: "multiplier",
                type: "uint8[][]",
            },
            {
                indexed: false,
                internalType: "uint16[]",
                name: "resourceCount",
                type: "uint16[]",
            },
            {
                indexed: false,
                internalType: "uint16[]",
                name: "resourceType",
                type: "uint16[]",
            },
            {
                indexed: false,
                internalType: "bool",
                name: "cidLock",
                type: "bool",
            },
        ],
        name: "CreateApp",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256[]",
                name: "appID",
                type: "uint256[]",
            },
            {
                indexed: false,
                internalType: "bytes32[]",
                name: "appName",
                type: "bytes32[]",
            },
            {
                indexed: false,
                internalType: "bytes[2][]",
                name: "path",
                type: "bytes[2][]",
            },
            {
                indexed: false,
                internalType: "uint256[][]",
                name: "subnetList",
                type: "uint256[][]",
            },
            {
                indexed: false,
                internalType: "uint8[][][]",
                name: "multiplier",
                type: "uint8[][][]",
            },
            {
                indexed: false,
                internalType: "uint16[][]",
                name: "resourceCount",
                type: "uint16[][]",
            },
            {
                indexed: false,
                internalType: "uint16[][]",
                name: "resourceType",
                type: "uint16[][]",
            },
            {
                indexed: false,
                internalType: "bool[]",
                name: "cidLock",
                type: "bool[]",
            },
        ],
        name: "CreateAppBatch",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "appID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "bytes32",
                name: "appName",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "uint256[]",
                name: "subnetList",
                type: "uint256[]",
            },
            {
                indexed: false,
                internalType: "uint8[][]",
                name: "currentReplicaList",
                type: "uint8[][]",
            },
            {
                indexed: false,
                internalType: "uint16[]",
                name: "resourceCount",
                type: "uint16[]",
            },
            {
                indexed: false,
                internalType: "uint16[]",
                name: "resourceType",
                type: "uint16[]",
            },
        ],
        name: "DeleteApp",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                components: [
                    {
                        internalType: "address",
                        name: "target",
                        type: "address",
                    },
                    {
                        internalType: "enum IDiamondWritableInternal.FacetCutAction",
                        name: "action",
                        type: "uint8",
                    },
                    {
                        internalType: "bytes4[]",
                        name: "selectors",
                        type: "bytes4[]",
                    },
                ],
                indexed: false,
                internalType: "struct IDiamondWritableInternal.FacetCut[]",
                name: "facetCuts",
                type: "tuple[]",
            },
            {
                indexed: false,
                internalType: "address",
                name: "target",
                type: "address",
            },
            {
                indexed: false,
                internalType: "bytes",
                name: "data",
                type: "bytes",
            },
        ],
        name: "DiamondCut",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "previousOwner",
                type: "address",
            },
            {
                indexed: true,
                internalType: "address",
                name: "newOwner",
                type: "address",
            },
        ],
        name: "OwnershipTransferred",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "balanceToAdd",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "appID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "bytes32",
                name: "appName",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "bytes[2]",
                name: "path",
                type: "bytes[2]",
            },
            {
                indexed: false,
                internalType: "uint256[]",
                name: "subnetList",
                type: "uint256[]",
            },
            {
                indexed: false,
                internalType: "uint8[][]",
                name: "multiplier",
                type: "uint8[][]",
            },
            {
                indexed: false,
                internalType: "uint16[]",
                name: "resourceCount",
                type: "uint16[]",
            },
            {
                indexed: false,
                internalType: "uint16[]",
                name: "resourceType",
                type: "uint16[]",
            },
        ],
        name: "UpdateApp",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "balanceToAdd",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256[]",
                name: "appID",
                type: "uint256[]",
            },
            {
                indexed: false,
                internalType: "bytes32[]",
                name: "appName",
                type: "bytes32[]",
            },
            {
                indexed: false,
                internalType: "bytes[2][]",
                name: "path",
                type: "bytes[2][]",
            },
            {
                indexed: false,
                internalType: "uint256[][]",
                name: "subnetList",
                type: "uint256[][]",
            },
            {
                indexed: false,
                internalType: "uint8[][][]",
                name: "multiplier",
                type: "uint8[][][]",
            },
            {
                indexed: false,
                internalType: "uint16[][]",
                name: "resourceCount",
                type: "uint16[][]",
            },
            {
                indexed: false,
                internalType: "uint16[][]",
                name: "resourceType",
                type: "uint16[][]",
            },
        ],
        name: "UpdateAppBatch",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "appID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "bytes[2]",
                name: "path",
                type: "bytes[2]",
            },
        ],
        name: "UpdatePath",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256[]",
                name: "appIDList",
                type: "uint256[]",
            },
            {
                indexed: false,
                internalType: "uint256[][]",
                name: "subnetList",
                type: "uint256[][]",
            },
            {
                indexed: false,
                internalType: "uint8[][][]",
                name: "multiplierList",
                type: "uint8[][][]",
            },
            {
                indexed: false,
                internalType: "uint16[][]",
                name: "resourceList",
                type: "uint16[][]",
            },
            {
                indexed: false,
                internalType: "uint16[][]",
                name: "resourceType",
                type: "uint16[][]",
            },
        ],
        name: "UpdateResource",
        type: "event",
    },
    { stateMutability: "payable", type: "fallback" },
    {
        inputs: [],
        name: "DEPLOYER",
        outputs: [{ internalType: "bytes32", name: "", type: "bytes32" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "Subscription",
        outputs: [
            {
                internalType: "contract ISubscription",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "SubscriptionBalance",
        outputs: [
            {
                internalType: "contract ISubscriptionBalance",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "acceptOwnership",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            { internalType: "uint256", name: "", type: "uint256" },
            { internalType: "uint256", name: "", type: "uint256" },
            { internalType: "uint256", name: "", type: "uint256" },
            { internalType: "uint256", name: "", type: "uint256" },
        ],
        name: "appCurrentReplica",
        outputs: [{ internalType: "uint8", name: "", type: "uint8" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            { internalType: "uint256", name: "", type: "uint256" },
            { internalType: "uint256", name: "", type: "uint256" },
        ],
        name: "appSubnetBitmap",
        outputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                components: [
                    {
                        internalType: "address",
                        name: "target",
                        type: "address",
                    },
                    {
                        internalType: "enum IDiamondWritableInternal.FacetCutAction",
                        name: "action",
                        type: "uint8",
                    },
                    {
                        internalType: "bytes4[]",
                        name: "selectors",
                        type: "bytes4[]",
                    },
                ],
                internalType: "struct IDiamondWritableInternal.FacetCut[]",
                name: "facetCuts",
                type: "tuple[]",
            },
            {
                internalType: "address",
                name: "target",
                type: "address",
            },
            { internalType: "bytes", name: "data", type: "bytes" },
        ],
        name: "diamondCut",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            { internalType: "uint256", name: "", type: "uint256" },
            { internalType: "uint256", name: "", type: "uint256" },
        ],
        name: "entries",
        outputs: [
            {
                internalType: "bytes32",
                name: "appName",
                type: "bytes32",
            },
            {
                internalType: "uint256",
                name: "timestamp",
                type: "uint256",
            },
            { internalType: "bool", name: "cidLock", type: "bool" },
            { internalType: "bool", name: "active", type: "bool" },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes4",
                name: "selector",
                type: "bytes4",
            },
        ],
        name: "facetAddress",
        outputs: [
            {
                internalType: "address",
                name: "facet",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "facetAddresses",
        outputs: [
            {
                internalType: "address[]",
                name: "addresses",
                type: "address[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "facet",
                type: "address",
            },
        ],
        name: "facetFunctionSelectors",
        outputs: [
            {
                internalType: "bytes4[]",
                name: "selectors",
                type: "bytes4[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "facets",
        outputs: [
            {
                components: [
                    {
                        internalType: "address",
                        name: "target",
                        type: "address",
                    },
                    {
                        internalType: "bytes4[]",
                        name: "selectors",
                        type: "bytes4[]",
                    },
                ],
                internalType: "struct IDiamondReadable.Facet[]",
                name: "diamondFacets",
                type: "tuple[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "getFallbackAddress",
        outputs: [
            {
                internalType: "address",
                name: "fallbackAddress",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        name: "lastAppID",
        outputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        name: "nftActiveSubnetCheck",
        outputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            { internalType: "uint256", name: "", type: "uint256" },
            { internalType: "uint256", name: "", type: "uint256" },
        ],
        name: "nftAllSubnets",
        outputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        name: "nftAppCount",
        outputs: [{ internalType: "uint16", name: "", type: "uint16" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            { internalType: "uint256", name: "", type: "uint256" },
            { internalType: "uint256", name: "", type: "uint256" },
        ],
        name: "nftSubnetEntry",
        outputs: [
            { internalType: "uint8", name: "entryID", type: "uint8" },
            { internalType: "uint8", name: "appCount", type: "uint8" },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "nomineeOwner",
        outputs: [{ internalType: "address", name: "", type: "address" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "owner",
        outputs: [{ internalType: "address", name: "", type: "address" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "fallbackAddress",
                type: "address",
            },
        ],
        name: "setFallbackAddress",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes4",
                name: "interfaceId",
                type: "bytes4",
            },
        ],
        name: "supportsInterface",
        outputs: [{ internalType: "bool", name: "", type: "bool" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "account",
                type: "address",
            },
        ],
        name: "transferOwnership",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    { stateMutability: "payable", type: "receive" },
];
exports.default = AppDeployment;
