// @decloudlabs/stk-v2/lib/ABI/SubnetCluster.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var SubnetCluster = [
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "address[]",
                name: "_whitelistAddresses",
                type: "address[]",
            },
        ],
        name: "addClusterToWhitelisted",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "contract IERC721Upgradeable",
                name: "nftContract",
                type: "address",
            },
        ],
        name: "addNFTContract",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "clusterID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "_weight",
                type: "uint256",
            },
        ],
        name: "approveListingCluster",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "clusterID",
                type: "uint256",
            },
            {
                internalType: "string",
                name: "newDNSIP",
                type: "string",
            },
        ],
        name: "changeClusterDNSIP",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "clusterID",
                type: "uint256",
            },
            {
                internalType: "string",
                name: "clusterName",
                type: "string",
            },
        ],
        name: "changeClusterName",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "_coolDownTimeSecs",
                type: "uint256",
            },
        ],
        name: "changeCoolDownTime",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "_REQD_STACK_FEES_FOR_SUBNET",
                type: "uint256",
            },
        ],
        name: "changeSubnetFees",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "_daoRate",
                type: "uint256",
            },
        ],
        name: "change_DAORate",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
        ],
        name: "checkMaxClustersJoined",
        outputs: [{ internalType: "bool", name: "", type: "bool" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "uint256[2]",
                name: "nftAndContract",
                type: "uint256[2]",
            },
            {
                internalType: "string",
                name: "_DNSIP",
                type: "string",
            },
            {
                internalType: "address",
                name: "walletAddress",
                type: "address",
            },
            {
                internalType: "address",
                name: "operatorAddress",
                type: "address",
            },
            {
                internalType: "string",
                name: "clusterName",
                type: "string",
            },
        ],
        name: "clusterSignUp",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "clusterID",
                type: "uint256",
            },
        ],
        name: "delistCluster",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "clusterID",
                type: "uint256",
            },
        ],
        name: "getClusterWalletAddress",
        outputs: [{ internalType: "address", name: "", type: "address" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            { internalType: "address", name: "user", type: "address" },
        ],
        name: "hasPermissionToClusterList",
        outputs: [{ internalType: "bool", name: "", type: "bool" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "clusterID",
                type: "uint256",
            },
        ],
        name: "isClusterListed",
        outputs: [{ internalType: "bool", name: "", type: "bool" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "pause",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "address",
                name: "_blacklistAddress",
                type: "address",
            },
        ],
        name: "removeClusterFromWhitelist",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "clusterID",
                type: "uint256",
            },
            {
                internalType: "address",
                name: "newOwnerAddress",
                type: "address",
            },
            {
                internalType: "address",
                name: "newWalletAddress",
                type: "address",
            },
            {
                internalType: "address",
                name: "newOperatorAddress",
                type: "address",
            },
        ],
        name: "transferClusterOwnership",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "unpause",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "clusterID",
                type: "uint256",
            },
        ],
        name: "withdrawClusterForDelistedSubnet",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "contractID",
                type: "uint256",
            },
            {
                internalType: "uint256[]",
                name: "nftIDList",
                type: "uint256[]",
            },
        ],
        name: "withdrawNFT",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "clusterID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "_amount",
                type: "uint256",
            },
        ],
        name: "withdrawStackFromClusterByDAO",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
        ],
        name: "activateSubnet",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "uint256[]",
                name: "computePriceList",
                type: "uint256[]",
            },
        ],
        name: "addSubnetComputePrices",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
        ],
        name: "applyChangedComputePrice",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            { internalType: "string", name: "dnsip", type: "string" },
        ],
        name: "changeSubnetDNSIP",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "address",
                name: "subnetLocalDAO",
                type: "address",
            },
        ],
        name: "changeSubnetLocalDAO",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "maxClusters",
                type: "uint256",
            },
        ],
        name: "changeSubnetMaxClusters",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "string",
                name: "subnetName",
                type: "string",
            },
        ],
        name: "changeSubnetName",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "uint256[]",
                name: "otherAttributes",
                type: "uint256[]",
            },
        ],
        name: "changeSubnetOtherAttributes",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "stackFeesReqd",
                type: "uint256",
            },
        ],
        name: "changeSubnetStackFeesReqd",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "subnetType",
                type: "uint256",
            },
        ],
        name: "changeSubnetType",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256[]",
                name: "subnetList",
                type: "uint256[]",
            },
        ],
        name: "checkSubnetStatus",
        outputs: [{ internalType: "uint8[]", name: "", type: "uint8[]" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256[2]",
                name: "nftAndContract",
                type: "uint256[2]",
            },
            {
                internalType: "address",
                name: "_subnetLocalDAO",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "_subnetType",
                type: "uint256",
            },
            {
                internalType: "bool",
                name: "_subnetStatusListed",
                type: "bool",
            },
            {
                internalType: "uint256[]",
                name: "computePrices",
                type: "uint256[]",
            },
            {
                internalType: "uint256[]",
                name: "_otherAttributes",
                type: "uint256[]",
            },
            {
                internalType: "uint256",
                name: "_maxClusters",
                type: "uint256",
            },
            {
                internalType: "address[]",
                name: "_whiteListedClusters",
                type: "address[]",
            },
            {
                internalType: "bytes",
                name: "publicKey",
                type: "bytes",
            },
            { internalType: "string", name: "dnsip", type: "string" },
            {
                internalType: "uint256",
                name: "_stackFeesReqd",
                type: "uint256",
            },
            {
                internalType: "string",
                name: "subnetName",
                type: "string",
            },
        ],
        name: "createSubnet",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
        ],
        name: "deactivateSubnet",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "getAllSubnetNamesAndIDs",
        outputs: [
            {
                components: [
                    {
                        internalType: "string",
                        name: "subnetName",
                        type: "string",
                    },
                    {
                        internalType: "uint256",
                        name: "subnetID",
                        type: "uint256",
                    },
                ],
                internalType: "struct SubnetClusterData.SubnetNameAndID[]",
                name: "subnetNameList",
                type: "tuple[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "_subnetId",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "_clusterId",
                type: "uint256",
            },
        ],
        name: "getClusterAttributes",
        outputs: [
            {
                components: [
                    {
                        internalType: "address",
                        name: "walletAddress",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "ownerAddress",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "operatorAddress",
                        type: "address",
                    },
                    {
                        internalType: "string",
                        name: "DNSIP",
                        type: "string",
                    },
                    {
                        internalType: "uint8",
                        name: "listed",
                        type: "uint8",
                    },
                    {
                        internalType: "uint256[]",
                        name: "lockedNFT",
                        type: "uint256[]",
                    },
                    {
                        internalType: "string",
                        name: "clusterName",
                        type: "string",
                    },
                ],
                internalType: "struct SubnetClusterData.Cluster",
                name: "",
                type: "tuple",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "uint16[]",
                name: "resourceType",
                type: "uint16[]",
            },
        ],
        name: "getComputePrices",
        outputs: [
            {
                internalType: "uint256[]",
                name: "computePrices",
                type: "uint256[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetId",
                type: "uint256",
            },
        ],
        name: "getSubnetAttributes",
        outputs: [
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "subnetType",
                        type: "uint256",
                    },
                    {
                        internalType: "uint8",
                        name: "subnetStatusListed",
                        type: "uint8",
                    },
                    {
                        internalType: "uint256[]",
                        name: "otherAttributes",
                        type: "uint256[]",
                    },
                    {
                        internalType: "uint256",
                        name: "maxClusters",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "stackFeesReqd",
                        type: "uint256",
                    },
                    {
                        internalType: "bytes32",
                        name: "SUBNET_DAO_ROLE",
                        type: "bytes32",
                    },
                    {
                        internalType: "bytes32",
                        name: "SUBNET_ATTR_ROLE",
                        type: "bytes32",
                    },
                    {
                        internalType: "bytes32",
                        name: "PRICE_ROLE",
                        type: "bytes32",
                    },
                    {
                        internalType: "bytes32",
                        name: "WHITELIST_ROLE",
                        type: "bytes32",
                    },
                    {
                        internalType: "bytes32",
                        name: "CLUSTER_LIST_ROLE",
                        type: "bytes32",
                    },
                    {
                        internalType: "bytes",
                        name: "publicKey",
                        type: "bytes",
                    },
                    {
                        internalType: "address",
                        name: "subnetLocalDAO",
                        type: "address",
                    },
                    {
                        internalType: "string",
                        name: "dnsip",
                        type: "string",
                    },
                    {
                        internalType: "string",
                        name: "subnetName",
                        type: "string",
                    },
                ],
                internalType: "struct SubnetClusterData.SubnetAttributes",
                name: "",
                type: "tuple",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "uint16[]",
                name: "typeList",
                type: "uint16[]",
            },
            {
                internalType: "uint256[]",
                name: "computePrices",
                type: "uint256[]",
            },
        ],
        name: "requestComputePriceChange",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "publicKey",
                type: "bytes",
            },
        ],
        name: "setSubnetPublicKey",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "contract ISubnetDAODistributor",
                name: "_SubnetDAODistributor",
                type: "address",
            },
        ],
        name: "set_SubnetDAODistributorContract",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "globalDAO",
                type: "address",
            },
        ],
        name: "set_globalDAO",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "CLUSTER_LIST_ROLE",
        outputs: [{ internalType: "bytes32", name: "", type: "bytes32" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "COOLDOWN_ROLE",
        outputs: [{ internalType: "bytes32", name: "", type: "bytes32" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "DEFAULT_ADMIN_ROLE",
        outputs: [{ internalType: "bytes32", name: "", type: "bytes32" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "DefaultWhitelistedClusterWeight",
        outputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "GLOBAL_DAO_ADDRESS",
        outputs: [{ internalType: "address", name: "", type: "address" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "MAX_ACTIVE_CLUSTERS",
        outputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "MAX_CLUSTER_WEIGHT",
        outputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "PAUSER_ROLE",
        outputs: [{ internalType: "bytes32", name: "", type: "bytes32" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "PRICE_ROLE",
        outputs: [{ internalType: "bytes32", name: "", type: "bytes32" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "REQD_STACK_FEES_FOR_SUBNET",
        outputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "SUBNET_ATTR_ROLE",
        outputs: [{ internalType: "bytes32", name: "", type: "bytes32" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "StackToken",
        outputs: [
            {
                internalType: "contract IERC20Upgradeable",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "SubnetDAODistributor",
        outputs: [
            {
                internalType: "contract ISubnetDAODistributor",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "WHITELIST_ROLE",
        outputs: [{ internalType: "bytes32", name: "", type: "bytes32" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "WITHDRAW_STACK_ROLE",
        outputs: [{ internalType: "bytes32", name: "", type: "bytes32" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "acceptOwnership",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        name: "activeClusterCount",
        outputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [{ internalType: "address", name: "", type: "address" }],
        name: "balanceOfStackLocked",
        outputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "coolDownTimeForPriceChange",
        outputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "daoRate",
        outputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        name: "darkMatterNFTList",
        outputs: [
            {
                internalType: "contract IERC721Upgradeable",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                components: [
                    {
                        internalType: "address",
                        name: "target",
                        type: "address",
                    },
                    {
                        internalType: "enum IDiamondWritableInternal.FacetCutAction",
                        name: "action",
                        type: "uint8",
                    },
                    {
                        internalType: "bytes4[]",
                        name: "selectors",
                        type: "bytes4[]",
                    },
                ],
                internalType: "struct IDiamondWritableInternal.FacetCut[]",
                name: "facetCuts",
                type: "tuple[]",
            },
            {
                internalType: "address",
                name: "target",
                type: "address",
            },
            { internalType: "bytes", name: "data", type: "bytes" },
        ],
        name: "diamondCut",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes4",
                name: "selector",
                type: "bytes4",
            },
        ],
        name: "facetAddress",
        outputs: [
            {
                internalType: "address",
                name: "facet",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "facetAddresses",
        outputs: [
            {
                internalType: "address[]",
                name: "addresses",
                type: "address[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "facet",
                type: "address",
            },
        ],
        name: "facetFunctionSelectors",
        outputs: [
            {
                internalType: "bytes4[]",
                name: "selectors",
                type: "bytes4[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "facets",
        outputs: [
            {
                components: [
                    {
                        internalType: "address",
                        name: "target",
                        type: "address",
                    },
                    {
                        internalType: "bytes4[]",
                        name: "selectors",
                        type: "bytes4[]",
                    },
                ],
                internalType: "struct IDiamondReadable.Facet[]",
                name: "diamondFacets",
                type: "tuple[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "getFallbackAddress",
        outputs: [
            {
                internalType: "address",
                name: "fallbackAddress",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [{ internalType: "bytes32", name: "role", type: "bytes32" }],
        name: "getRoleAdmin",
        outputs: [{ internalType: "bytes32", name: "", type: "bytes32" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "role",
                type: "bytes32",
            },
            {
                internalType: "address",
                name: "account",
                type: "address",
            },
        ],
        name: "grantRole",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "role",
                type: "bytes32",
            },
            {
                internalType: "address",
                name: "account",
                type: "address",
            },
        ],
        name: "hasRole",
        outputs: [{ internalType: "bool", name: "", type: "bool" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "nomineeOwner",
        outputs: [{ internalType: "address", name: "", type: "address" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "owner",
        outputs: [{ internalType: "address", name: "", type: "address" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "paused",
        outputs: [{ internalType: "bool", name: "status", type: "bool" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [{ internalType: "bytes32", name: "role", type: "bytes32" }],
        name: "renounceRole",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "role",
                type: "bytes32",
            },
            {
                internalType: "address",
                name: "account",
                type: "address",
            },
        ],
        name: "revokeRole",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "fallbackAddress",
                type: "address",
            },
        ],
        name: "setFallbackAddress",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        name: "subnetAttributes",
        outputs: [
            {
                internalType: "uint256",
                name: "subnetType",
                type: "uint256",
            },
            {
                internalType: "uint8",
                name: "subnetStatusListed",
                type: "uint8",
            },
            {
                internalType: "uint256",
                name: "maxClusters",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "stackFeesReqd",
                type: "uint256",
            },
            {
                internalType: "bytes32",
                name: "SUBNET_DAO_ROLE",
                type: "bytes32",
            },
            {
                internalType: "bytes32",
                name: "SUBNET_ATTR_ROLE",
                type: "bytes32",
            },
            {
                internalType: "bytes32",
                name: "PRICE_ROLE",
                type: "bytes32",
            },
            {
                internalType: "bytes32",
                name: "WHITELIST_ROLE",
                type: "bytes32",
            },
            {
                internalType: "bytes32",
                name: "CLUSTER_LIST_ROLE",
                type: "bytes32",
            },
            {
                internalType: "bytes",
                name: "publicKey",
                type: "bytes",
            },
            {
                internalType: "address",
                name: "subnetLocalDAO",
                type: "address",
            },
            { internalType: "string", name: "dnsip", type: "string" },
            {
                internalType: "string",
                name: "subnetName",
                type: "string",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            { internalType: "uint256", name: "", type: "uint256" },
            { internalType: "uint256", name: "", type: "uint256" },
        ],
        name: "subnetClusters",
        outputs: [
            {
                internalType: "address",
                name: "walletAddress",
                type: "address",
            },
            {
                internalType: "address",
                name: "ownerAddress",
                type: "address",
            },
            {
                internalType: "address",
                name: "operatorAddress",
                type: "address",
            },
            { internalType: "string", name: "DNSIP", type: "string" },
            { internalType: "uint8", name: "listed", type: "uint8" },
            {
                internalType: "string",
                name: "clusterName",
                type: "string",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes4",
                name: "interfaceId",
                type: "bytes4",
            },
        ],
        name: "supportsInterface",
        outputs: [{ internalType: "bool", name: "", type: "bool" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        name: "totalClustersSigned",
        outputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "totalSubnets",
        outputs: [{ internalType: "uint256", name: "", type: "uint256" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "account",
                type: "address",
            },
        ],
        name: "transferOwnership",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            { internalType: "uint256", name: "", type: "uint256" },
            { internalType: "address", name: "", type: "address" },
        ],
        name: "whiteListedClusters",
        outputs: [{ internalType: "bool", name: "", type: "bool" }],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "contract IERC20Upgradeable",
                name: "_StackToken",
                type: "address",
            },
            {
                internalType: "address",
                name: "_GlobalDAO",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "_coolDownTimeForPriceChange",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "_daoRate",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "_REQD_STACK_FEES_FOR_SUBNET",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "_DefaultWhitelistedClusterWeight",
                type: "uint256",
            },
        ],
        stateMutability: "nonpayable",
        type: "constructor",
    },
    {
        inputs: [],
        name: "DiamondWritable__InvalidInitializationParameters",
        type: "error",
    },
    {
        inputs: [],
        name: "DiamondWritable__RemoveTargetNotZeroAddress",
        type: "error",
    },
    {
        inputs: [],
        name: "DiamondWritable__ReplaceTargetIsIdentical",
        type: "error",
    },
    {
        inputs: [],
        name: "DiamondWritable__SelectorAlreadyAdded",
        type: "error",
    },
    {
        inputs: [],
        name: "DiamondWritable__SelectorIsImmutable",
        type: "error",
    },
    {
        inputs: [],
        name: "DiamondWritable__SelectorNotFound",
        type: "error",
    },
    {
        inputs: [],
        name: "DiamondWritable__SelectorNotSpecified",
        type: "error",
    },
    {
        inputs: [],
        name: "DiamondWritable__TargetHasNoCode",
        type: "error",
    },
    {
        inputs: [],
        name: "ERC165Base__InvalidInterfaceId",
        type: "error",
    },
    { inputs: [], name: "Ownable__NotOwner", type: "error" },
    {
        inputs: [],
        name: "Ownable__NotTransitiveOwner",
        type: "error",
    },
    { inputs: [], name: "Pausable__NotPaused", type: "error" },
    { inputs: [], name: "Pausable__Paused", type: "error" },
    {
        inputs: [],
        name: "Proxy__ImplementationIsNotContract",
        type: "error",
    },
    {
        inputs: [],
        name: "SafeOwnable__NotNomineeOwner",
        type: "error",
    },
    {
        inputs: [],
        name: "UintUtils__InsufficientHexLength",
        type: "error",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "contract IERC721Upgradeable",
                name: "nftContract",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "contractID",
                type: "uint256",
            },
        ],
        name: "AddedDarkMatterNFTContract",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "address[]",
                name: "whitelistAddresses",
                type: "address[]",
            },
        ],
        name: "AddedToWhitelistCluster",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint16[]",
                name: "typeList",
                type: "uint16[]",
            },
            {
                indexed: false,
                internalType: "uint256[]",
                name: "computePrices",
                type: "uint256[]",
            },
        ],
        name: "AppliedChangedClusterPrice",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "clusterID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "string",
                name: "clusterName",
                type: "string",
            },
        ],
        name: "ChangedClusterName",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "coolDownTimeSecs",
                type: "uint256",
            },
        ],
        name: "ChangedCoolDownForPriceChange",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "clusterId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "string",
                name: "newDNSIP",
                type: "string",
            },
        ],
        name: "ChangedDNSIP",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "clusterId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint8",
                name: "allowStatus",
                type: "uint8",
            },
        ],
        name: "ChangedListingCluster",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "DarkMatterNFT",
                type: "address",
            },
        ],
        name: "ChangedNFTAddress",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "clusterId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "string",
                name: "DNS_IP",
                type: "string",
            },
            {
                indexed: false,
                internalType: "address",
                name: "walletAddress",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "ownerAddress",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "operatorAddress",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: false,
                internalType: "string",
                name: "clusterName",
                type: "string",
            },
        ],
        name: "ClusterSignedUp",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "daoRate",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "_daoRate",
                type: "uint256",
            },
        ],
        name: "DAORateChanged",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                components: [
                    {
                        internalType: "address",
                        name: "target",
                        type: "address",
                    },
                    {
                        internalType: "enum IDiamondWritableInternal.FacetCutAction",
                        name: "action",
                        type: "uint8",
                    },
                    {
                        internalType: "bytes4[]",
                        name: "selectors",
                        type: "bytes4[]",
                    },
                ],
                indexed: false,
                internalType: "struct IDiamondWritableInternal.FacetCut[]",
                name: "facetCuts",
                type: "tuple[]",
            },
            {
                indexed: false,
                internalType: "address",
                name: "target",
                type: "address",
            },
            {
                indexed: false,
                internalType: "bytes",
                name: "data",
                type: "bytes",
            },
        ],
        name: "DiamondCut",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "oldDAO",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "newDAO",
                type: "address",
            },
        ],
        name: "GlobalDAOChanged",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "userAddress",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "clusterId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256[2]",
                name: "nftIDAndContract",
                type: "uint256[2]",
            },
        ],
        name: "NFTLockedForCluster",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "userAddress",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256[2]",
                name: "nftIDAndContract",
                type: "uint256[2]",
            },
        ],
        name: "NFTLockedForSubnet",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "previousOwner",
                type: "address",
            },
            {
                indexed: true,
                internalType: "address",
                name: "newOwner",
                type: "address",
            },
        ],
        name: "OwnershipTransferred",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "account",
                type: "address",
            },
        ],
        name: "Paused",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "address",
                name: "blacklistAddress",
                type: "address",
            },
        ],
        name: "RemovedWhitelistCluster",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint16[]",
                name: "typeList",
                type: "uint16[]",
            },
            {
                indexed: false,
                internalType: "uint256[]",
                name: "computePrices",
                type: "uint256[]",
            },
        ],
        name: "RequestedClusterPriceChange",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetId",
                type: "uint256",
            },
        ],
        name: "ResetWhitelistCluster",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "bytes32",
                name: "role",
                type: "bytes32",
            },
            {
                indexed: true,
                internalType: "bytes32",
                name: "previousAdminRole",
                type: "bytes32",
            },
            {
                indexed: true,
                internalType: "bytes32",
                name: "newAdminRole",
                type: "bytes32",
            },
        ],
        name: "RoleAdminChanged",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "bytes32",
                name: "role",
                type: "bytes32",
            },
            {
                indexed: true,
                internalType: "address",
                name: "account",
                type: "address",
            },
            {
                indexed: true,
                internalType: "address",
                name: "sender",
                type: "address",
            },
        ],
        name: "RoleGranted",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "bytes32",
                name: "role",
                type: "bytes32",
            },
            {
                indexed: true,
                internalType: "address",
                name: "account",
                type: "address",
            },
            {
                indexed: true,
                internalType: "address",
                name: "sender",
                type: "address",
            },
        ],
        name: "RoleRevoked",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetId",
                type: "uint256",
            },
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "subnetType",
                        type: "uint256",
                    },
                    {
                        internalType: "uint8",
                        name: "subnetStatusListed",
                        type: "uint8",
                    },
                    {
                        internalType: "uint256[]",
                        name: "otherAttributes",
                        type: "uint256[]",
                    },
                    {
                        internalType: "uint256",
                        name: "maxClusters",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "stackFeesReqd",
                        type: "uint256",
                    },
                    {
                        internalType: "bytes32",
                        name: "SUBNET_DAO_ROLE",
                        type: "bytes32",
                    },
                    {
                        internalType: "bytes32",
                        name: "SUBNET_ATTR_ROLE",
                        type: "bytes32",
                    },
                    {
                        internalType: "bytes32",
                        name: "PRICE_ROLE",
                        type: "bytes32",
                    },
                    {
                        internalType: "bytes32",
                        name: "WHITELIST_ROLE",
                        type: "bytes32",
                    },
                    {
                        internalType: "bytes32",
                        name: "CLUSTER_LIST_ROLE",
                        type: "bytes32",
                    },
                    {
                        internalType: "bytes",
                        name: "publicKey",
                        type: "bytes",
                    },
                    {
                        internalType: "address",
                        name: "subnetLocalDAO",
                        type: "address",
                    },
                    {
                        internalType: "string",
                        name: "dnsip",
                        type: "string",
                    },
                    {
                        internalType: "string",
                        name: "subnetName",
                        type: "string",
                    },
                ],
                indexed: false,
                internalType: "struct SubnetClusterData.SubnetAttributes",
                name: "subnetAttributes",
                type: "tuple",
            },
            {
                indexed: false,
                internalType: "address",
                name: "sender",
                type: "address",
            },
        ],
        name: "SubnetAttributesChanged",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "address",
                name: "subnetLocalDAO",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetType",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint8",
                name: "subnetStatusListed",
                type: "uint8",
            },
            {
                indexed: false,
                internalType: "uint256[]",
                name: "otherAttributes",
                type: "uint256[]",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "maxClusters",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "address[]",
                name: "whiteListedClusters",
                type: "address[]",
            },
            {
                indexed: false,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: false,
                internalType: "string",
                name: "dnsip",
                type: "string",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "stackFeesReqd",
                type: "uint256",
            },
        ],
        name: "SubnetCreated",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "bytes",
                name: "publicKey",
                type: "bytes",
            },
        ],
        name: "SubnetPublicKeyChanged",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "clusterId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "newOwnerAddress",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "newWalletAddress",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "newOperatorAddress",
                type: "address",
            },
        ],
        name: "TransferredClusterOwnership",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "account",
                type: "address",
            },
        ],
        name: "Unpaused",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "clusterId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256[]",
                name: "nftAndContract",
                type: "uint256[]",
            },
        ],
        name: "WithdrawnNFTFromCluster",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256[]",
                name: "nftIds",
                type: "uint256[]",
            },
        ],
        name: "WithdrawnNFTs",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "clusterId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "stackTokens",
                type: "uint256",
            },
        ],
        name: "WithdrawnStackFromCluster",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "subnetId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "clusterId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
        ],
        name: "WithdrawnStackFromClusterByDAO",
        type: "event",
    },
    { stateMutability: "payable", type: "fallback" },
    { stateMutability: "payable", type: "receive" },
];
exports.default = SubnetCluster;
