// @decloudlabs/stk-v2/lib/ABI/Subscription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Subscription = [
    {
        inputs: [],
        stateMutability: "nonpayable",
        type: "constructor",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "address",
                name: "supportAddress",
                type: "address",
            },
        ],
        name: "AppliedSupportChange",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "prev_limit",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "new_limit",
                type: "uint256",
            },
        ],
        name: "ChangedSupportAddressCooldown",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "prev_limit",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "new_limit",
                type: "uint256",
            },
        ],
        name: "ChangedSupportAddressNotice",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint8",
                name: "version",
                type: "uint8",
            },
        ],
        name: "Initialized",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "address",
                name: "licenseAddress",
                type: "address",
            },
        ],
        name: "LicenseAdded",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "address",
                name: "supportAddress",
                type: "address",
            },
        ],
        name: "NFTSupportAddressChanged",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256[]",
                name: "supportFactor",
                type: "uint256[]",
            },
        ],
        name: "NFTSupportFactorChanged",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "account",
                type: "address",
            },
        ],
        name: "Paused",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "platformAddress",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "platformPercentage",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "discountPercentage",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "referralPercentage",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "referralDuration",
                type: "uint256",
            },
        ],
        name: "PlatformAddressAdded",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "address",
                name: "referralAddress",
                type: "address",
            },
        ],
        name: "ReferralAdded",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "address",
                name: "newSupportAddress",
                type: "address",
            },
        ],
        name: "RequestedSupportChange",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "bytes32",
                name: "role",
                type: "bytes32",
            },
            {
                indexed: true,
                internalType: "bytes32",
                name: "previousAdminRole",
                type: "bytes32",
            },
            {
                indexed: true,
                internalType: "bytes32",
                name: "newAdminRole",
                type: "bytes32",
            },
        ],
        name: "RoleAdminChanged",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "bytes32",
                name: "role",
                type: "bytes32",
            },
            {
                indexed: true,
                internalType: "address",
                name: "account",
                type: "address",
            },
            {
                indexed: true,
                internalType: "address",
                name: "sender",
                type: "address",
            },
        ],
        name: "RoleGranted",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "bytes32",
                name: "role",
                type: "bytes32",
            },
            {
                indexed: true,
                internalType: "address",
                name: "account",
                type: "address",
            },
            {
                indexed: true,
                internalType: "address",
                name: "sender",
                type: "address",
            },
        ],
        name: "RoleRevoked",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "supportAddress",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256[]",
                name: "supportFactor",
                type: "uint256[]",
            },
        ],
        name: "SetSupportFactorForNFT",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "address[]",
                name: "factorAddressList",
                type: "address[]",
            },
            {
                indexed: false,
                internalType: "uint256[]",
                name: "licenseFactor",
                type: "uint256[]",
            },
            {
                indexed: false,
                internalType: "uint256[]",
                name: "supportFactor",
                type: "uint256[]",
            },
        ],
        name: "Subscribed",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "supportAddress",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256[]",
                name: "supportFactor",
                type: "uint256[]",
            },
        ],
        name: "SupportAddressAdded",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "account",
                type: "address",
            },
        ],
        name: "Unpaused",
        type: "event",
    },
    {
        inputs: [],
        name: "ADDR_LIMIT",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "ApplicationNFT",
        outputs: [
            {
                internalType: "contract IApplicationNFT",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "BRIDGE_ROLE",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "BalanceCalculator",
        outputs: [
            {
                internalType: "contract IBalanceCalculator",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "CHANGE_COMPUTE_ROLE",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "CHANGE_REQUEST_DURATION",
        outputs: [
            {
                internalType: "uint256",
                name: "supportAddressNoticeDuration",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "supportAddressCooldownDuration",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "CHANGE_SUBSCRIPTION_ROLE",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "DEFAULT_ADMIN_ROLE",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "GLOBAL_DAO_ADDRESS",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "GLOBAL_SUPPORT_ADDRESS",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "LICENSE_ADDR_ID",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "PAUSER_ROLE",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "PLATFORM_ADDR_ID",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "REFERRAL_ADDR_ID",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "SUBSCRIBE_ROLE",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "SUPPORT_ADDR_ID",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "SubnetCluster",
        outputs: [
            {
                internalType: "contract ISubnetCluster",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "SubscriptionBalance",
        outputs: [
            {
                internalType: "contract ISubscriptionBalance",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "WITHDRAW_CREDITS_ROLE",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "XCT",
        outputs: [
            {
                internalType: "contract IERC20Upgradeable",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "platformAddress",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "platformPercentage",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "discountPercentage",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "referralPercentage",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "referralDuration",
                type: "uint256",
            },
        ],
        name: "addPlatformAddress",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "nftOwner",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "address",
                name: "referralAddress",
                type: "address",
            },
        ],
        name: "addReferralAddress",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "supportAddress",
                type: "address",
            },
            {
                internalType: "uint256[]",
                name: "supportFactor",
                type: "uint256[]",
            },
        ],
        name: "addSupportAddress",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "newGlobalDAO",
                type: "address",
            },
        ],
        name: "admin_changeGlobalDAO",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "supportAddressCooldownDuration",
                type: "uint256",
            },
        ],
        name: "admin_changeSupportAddressCooldown",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "supportAddressNoticeDuration",
                type: "uint256",
            },
        ],
        name: "admin_changeSupportAddressNotice",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "nftOwner",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "applySupportChange",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "nftOwner",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "approveNewSupportFactor",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "caller",
                type: "address",
            },
            {
                internalType: "address",
                name: "newLicenseAddress",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "changeLicenseAddress",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "bridge",
                type: "address",
            },
        ],
        name: "checkBridgeRole",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "roleName",
                type: "string",
            },
        ],
        name: "getBytes32OfRole",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "pure",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "getCreateTime",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "getLicenseFactor",
        outputs: [
            {
                internalType: "uint256[]",
                name: "",
                type: "uint256[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "factorID",
                type: "uint256",
            },
        ],
        name: "getNFTFactorAddress",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "getNFTSubscription",
        outputs: [
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "createTime",
                        type: "uint256",
                    },
                ],
                internalType: "struct Subscription.NFTAttribute",
                name: "nftAttribute",
                type: "tuple",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "platformAddress",
                type: "address",
            },
        ],
        name: "getPlatformFactors",
        outputs: [
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "platformPercentage",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "discountPercentage",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "referralPercentage",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "referralExpiryDuration",
                        type: "uint256",
                    },
                    {
                        internalType: "bool",
                        name: "active",
                        type: "bool",
                    },
                ],
                internalType: "struct Subscription.PlatformAddress",
                name: "",
                type: "tuple",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "role",
                type: "bytes32",
            },
        ],
        name: "getRoleAdmin",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "getSupportFactor",
        outputs: [
            {
                internalType: "uint256[]",
                name: "",
                type: "uint256[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "supportAddress",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "getSupportFeesForNFT",
        outputs: [
            {
                internalType: "uint256[]",
                name: "supportFactor",
                type: "uint256[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "role",
                type: "bytes32",
            },
            {
                internalType: "address",
                name: "account",
                type: "address",
            },
        ],
        name: "grantRole",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "role",
                type: "bytes32",
            },
            {
                internalType: "address",
                name: "account",
                type: "address",
            },
        ],
        name: "hasRole",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "_GlobalDAO",
                type: "address",
            },
            {
                internalType: "address",
                name: "_globalSupportAddress",
                type: "address",
            },
            {
                internalType: "uint256[]",
                name: "_globalSupportFactor",
                type: "uint256[]",
            },
            {
                internalType: "contract ISubnetCluster",
                name: "_SubnetCluster",
                type: "address",
            },
            {
                internalType: "contract IApplicationNFT",
                name: "_ApplicationNFT",
                type: "address",
            },
            {
                internalType: "contract ISubscriptionBalance",
                name: "_SubscriptionBalance",
                type: "address",
            },
            {
                internalType: "contract IBalanceCalculator",
                name: "_BalanceCalculator",
                type: "address",
            },
            {
                internalType: "contract IERC20Upgradeable",
                name: "_XCT",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "_REQD_NOTICE_TIME_S_PROVIDER",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "_REQD_COOLDOWN_S_PROVIDER",
                type: "uint256",
            },
        ],
        name: "initialize",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "isBridgeRole",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        name: "nftSubscription",
        outputs: [
            {
                internalType: "uint256",
                name: "createTime",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "pause",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "paused",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        name: "platformAddressList",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        name: "platformAddressMap",
        outputs: [
            {
                internalType: "uint256",
                name: "platformPercentage",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "discountPercentage",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "referralPercentage",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "referralExpiryDuration",
                type: "uint256",
            },
            {
                internalType: "bool",
                name: "active",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "role",
                type: "bytes32",
            },
            {
                internalType: "address",
                name: "account",
                type: "address",
            },
        ],
        name: "renounceRole",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        name: "requestChangeMap",
        outputs: [
            {
                internalType: "uint256",
                name: "SupportAddressRequestTime",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "SupportAddressApplyTime",
                type: "uint256",
            },
            {
                internalType: "address",
                name: "supportAddress",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "nftOwner",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "address",
                name: "newSupportAddress",
                type: "address",
            },
        ],
        name: "requestSupportChange",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "role",
                type: "bytes32",
            },
            {
                internalType: "address",
                name: "account",
                type: "address",
            },
        ],
        name: "revokeRole",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "supportAddress",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "uint256[]",
                name: "supportFactor",
                type: "uint256[]",
            },
        ],
        name: "setSupportFactorForNFT",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "address[]",
                name: "addressList",
                type: "address[]",
            },
            {
                internalType: "uint256[]",
                name: "licenseFactor",
                type: "uint256[]",
            },
        ],
        name: "subscribe",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        name: "supportAddressList",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        name: "supportAddressToNFT",
        outputs: [
            {
                internalType: "bool",
                name: "active",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes4",
                name: "interfaceId",
                type: "bytes4",
            },
        ],
        name: "supportsInterface",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "unpause",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
exports.default = Subscription;
