// @decloudlabs/stk-v2/lib/ABI/SubscriptionBalance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var SubscriptionBalance = [
    {
        inputs: [],
        stateMutability: "nonpayable",
        type: "constructor",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "NFTId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "balanceType",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "bal",
                type: "uint256",
            },
        ],
        name: "BalanceAdded",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "BalanceUpdate",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "NFTId",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "balanceType",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "bal",
                type: "uint256",
            },
        ],
        name: "BalanceWithdrawn",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "oldMinDuration",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "oldMaxDuration",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "newMinDuration",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "newMaxDuration",
                type: "uint256",
            },
        ],
        name: "ChangedMaxExpiryDuration",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "DistributeAccumulatedRevenue",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint8",
                name: "version",
                type: "uint8",
            },
        ],
        name: "Initialized",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "previousOwner",
                type: "address",
            },
            {
                indexed: true,
                internalType: "address",
                name: "newOwner",
                type: "address",
            },
        ],
        name: "OwnershipTransferred",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "account",
                type: "address",
            },
        ],
        name: "Paused",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "benficiary",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "bal",
                type: "uint256",
            },
        ],
        name: "ReceivedRevenue",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "SubnetBalanceUpdate",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "account",
                type: "address",
            },
        ],
        name: "Unpaused",
        type: "event",
    },
    {
        inputs: [],
        name: "AppDeployment",
        outputs: [
            {
                internalType: "contract IAppDeployment",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "ApplicationNFT",
        outputs: [
            {
                internalType: "contract IApplicationNFT",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "BILLING_MANAGER_ROLE",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "BalanceCalculator",
        outputs: [
            {
                internalType: "contract IBalanceCalculator",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "LinkContract",
        outputs: [
            {
                internalType: "contract ILinkContract",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "SubnetCluster",
        outputs: [
            {
                internalType: "contract ISubnetCluster",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "SubscriptionContract",
        outputs: [
            {
                internalType: "contract ISubscription",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "XCTToken",
        outputs: [
            {
                internalType: "contract IERC20Upgradeable",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "nftOwner",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "balanceToAdd",
                type: "uint256",
            },
        ],
        name: "addBalance",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "_balanceToAdd",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "_expiryUnixTimestamp",
                type: "uint256",
            },
        ],
        name: "addBalanceAsCredit",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "_balanceToAdd",
                type: "uint256",
            },
        ],
        name: "addBalanceAsExternalDeposit",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "nftOwner",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "balanceToAdd",
                type: "uint256",
            },
        ],
        name: "addBalanceWithoutUpdate",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "account",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "balance",
                type: "uint256",
            },
        ],
        name: "addRevBalance",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        name: "balanceOfRev",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "checkBalanceLeft",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        name: "creditsExpiry",
        outputs: [
            {
                internalType: "uint256",
                name: "expiryTimestamp",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "amountAdded",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "distributeAccumulatedRevenue",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "getBalanceEndTime",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "getLastUpdateTime",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "contract ISubnetCluster",
                name: "_SubnetCluster",
                type: "address",
            },
            {
                internalType: "contract IApplicationNFT",
                name: "_ApplicationNFT",
                type: "address",
            },
            {
                internalType: "contract IERC20Upgradeable",
                name: "_XCTToken",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "_MIN_CREDIT_EXPIRY_DURATION",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "_MAX_CREDIT_EXPIRY_DURATION",
                type: "uint256",
            },
        ],
        name: "initialize",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "isBridgeRole",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        name: "nftAccumCost",
        outputs: [
            {
                internalType: "uint256",
                name: "accumCost",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "accumDuration",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        name: "nftBalances",
        outputs: [
            {
                internalType: "uint256",
                name: "lastBalanceUpdateTime",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "owner",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "pause",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "paused",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "prevBalances",
        outputs: [
            {
                internalType: "uint256[3]",
                name: "",
                type: "uint256[3]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "receiveRevenue",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "_userAddress",
                type: "address",
            },
        ],
        name: "receiveRevenueForAddress",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "renounceOwnership",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "s_GlobalDAOAddress",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "s_GlobalDAORate",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "_AppDeployment",
                type: "address",
            },
        ],
        name: "setAppDeployment",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "_BalanceCalculator",
                type: "address",
            },
        ],
        name: "setBalanceCalculator",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "_MIN_CREDIT_EXPIRY_DURATION",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "_MAX_CREDIT_EXPIRY_DURATION",
                type: "uint256",
            },
        ],
        name: "setExpiryDuration",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "contract ILinkContract",
                name: "_newLinkContract",
                type: "address",
            },
        ],
        name: "setLinkContract",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "_SubnetDAODistributor",
                type: "address",
            },
        ],
        name: "setSubnetDAODistributor",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "_Subscription",
                type: "address",
            },
        ],
        name: "setSubscriptionContract",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "subscribeNew",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "totalPrevBalance",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "newOwner",
                type: "address",
            },
        ],
        name: "transferOwnership",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "unpause",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "updateBalance",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "updateBalanceImmediate",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "nftOwner",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
        ],
        name: "withdrawAllOwnerBalance",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "nftOwner",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "_bal",
                type: "uint256",
            },
        ],
        name: "withdrawBalance",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "nftOwner",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "_bal",
                type: "uint256",
            },
            {
                internalType: "address",
                name: "customNFTContract",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "customNFTID",
                type: "uint256",
            },
        ],
        name: "withdrawBalanceLinked",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "nftID",
                type: "uint256",
            },
            {
                internalType: "address",
                name: "_to",
                type: "address",
            },
        ],
        name: "withdrawCreditsForNFT",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
exports.default = SubscriptionBalance;
