// @tria-sdk/web/dist/utils/chains/evm/utils/constants.js
export const constants = {
    evm: {
        ethereum: {
            chainId: 1,
            rpcUrl: 'https://eth-mainnet.g.alchemy.com/v2/z3iwhsUztqUGYH2fSuAP9uUA-0sBMIpC',
        },
        optimism: {
            chainId: 10,
            rpcUrl: 'https://optimism-mainnet.public.blastapi.io',
        },
        polygon: {
            chainId: 137,
            rpcUrl: 'https://polygon.llamarpc.com',
        },
        binance: {
            chainId: 56,
            rpcUrl: 'https://bsc.publicnode.com',
        },
        arbitrum: {
            chainId: 42161,
            rpcUrl: 'https://arbitrum-one.publicnode.com',
        },
        celo: {
            chainId: 42220,
            rpcUrl: 'https://rpc.ankr.com/celo',
        },
        avalanche: {
            chainId: 43114,
            rpcUrl: 'https://api.avax.network/ext/bc/C/rpc',
        },
        fantom: {
            chainId: 250,
            rpcUrl: 'https://rpc2.fantom.network',
        },
        gnosis: {
            chainId: 100,
            rpcUrl: 'https://rpc.ankr.com/gnosis',
        },
        sepolia: {
            chainId: 11155111,
            rpcUrl: 'https://eth-sepolia.g.alchemy.com/v2/z3iwhsUztqUGYH2fSuAP9uUA-0sBMIpC',
        },
        okx: {
            chainId: 66,
            rpcUrl: 'https://exchainrpc.okex.org',
        },
        fuse: {
            chainId: 122,
            rpcUrl: 'https://rpc.fuse.io/',
        },
    },
};
