// @decloudlabs/stk-v2/lib/types/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CRUD_APP_STAGE = exports.RESOURCE_CATEGORY = exports.RESTYPE_ID_TO_NAME_MAP = exports.RESTYPE_NAME_TO_ID_MAP = void 0;
var RESTYPE_NAME_TO_ID_MAP;
(function (RESTYPE_NAME_TO_ID_MAP) {
    RESTYPE_NAME_TO_ID_MAP[RESTYPE_NAME_TO_ID_MAP["cpuStandard"] = 0] = "cpuStandard";
    RESTYPE_NAME_TO_ID_MAP[RESTYPE_NAME_TO_ID_MAP["cpuIntensive"] = 1] = "cpuIntensive";
    RESTYPE_NAME_TO_ID_MAP[RESTYPE_NAME_TO_ID_MAP["gpuStandard"] = 2] = "gpuStandard";
    RESTYPE_NAME_TO_ID_MAP[RESTYPE_NAME_TO_ID_MAP["storage"] = 3] = "storage";
    RESTYPE_NAME_TO_ID_MAP[RESTYPE_NAME_TO_ID_MAP["bandwidth"] = 4] = "bandwidth";
    RESTYPE_NAME_TO_ID_MAP[RESTYPE_NAME_TO_ID_MAP["ipfsUpload"] = 5] = "ipfsUpload";
})(RESTYPE_NAME_TO_ID_MAP || (exports.RESTYPE_NAME_TO_ID_MAP = RESTYPE_NAME_TO_ID_MAP = {}));
var RESTYPE_ID_TO_NAME_MAP;
(function (RESTYPE_ID_TO_NAME_MAP) {
    RESTYPE_ID_TO_NAME_MAP["cpuStandard"] = "cpuStandard";
    RESTYPE_ID_TO_NAME_MAP["cpuIntensive"] = "cpuIntensive";
    RESTYPE_ID_TO_NAME_MAP["gpuStandard"] = "gpuStandard";
    RESTYPE_ID_TO_NAME_MAP["storage"] = "storage";
    RESTYPE_ID_TO_NAME_MAP["bandwidth"] = "bandwidth";
    RESTYPE_ID_TO_NAME_MAP["ipfsUpload"] = "ipfsUpload";
})(RESTYPE_ID_TO_NAME_MAP || (exports.RESTYPE_ID_TO_NAME_MAP = RESTYPE_ID_TO_NAME_MAP = {}));
exports.RESOURCE_CATEGORY = {
    cpuType: [0, 1, 2],
    storageType: [3],
    bandwidthType: [4],
    fileType: [5],
};
var CRUD_APP_STAGE;
(function (CRUD_APP_STAGE) {
    CRUD_APP_STAGE["STARTING_CREATE_APP"] = "STARTING_CREATE_APP";
    CRUD_APP_STAGE["STARTING_UPDATE_APP"] = "STARTING_UPDATE_APP";
    CRUD_APP_STAGE["STARTING_DELETE_APP"] = "STARTING_DELETE_APP";
    CRUD_APP_STAGE["VALIDATE_CONTRACT_APP"] = "VALIDATE_CONTRACT_APP";
    CRUD_APP_STAGE["CREATING_APP"] = "CREATING_APP";
    CRUD_APP_STAGE["UPDATING_APP"] = "UPDATING_APP";
    CRUD_APP_STAGE["UPDATING_CID"] = "UPDATING_CID";
    CRUD_APP_STAGE["DELETING_APP"] = "DELETING_APP";
    CRUD_APP_STAGE["APPROVE_BALANCE_WITHDRAW"] = "APPROVE_BALANCE_WITHDRAW";
    CRUD_APP_STAGE["SUBSCRIBING_CREATING_APP"] = "SUBSCRIBING_CREATING_APP";
    CRUD_APP_STAGE["SAVING_CONTRACT_APP_LOCAL"] = "SAVING_CONTRACT_APP_LOCAL";
    CRUD_APP_STAGE["SAVING_SUBSCRIPTION_LOCAL"] = "SAVING_SUBSCRIPTION_LOCAL";
    CRUD_APP_STAGE["DELETE_CONTRACT_APP_LOCAL"] = "DELETE_CONTRACT_APP_LOCAL";
    CRUD_APP_STAGE["ENCRYPTING_APP"] = "ENCRYPTING_APP";
    CRUD_APP_STAGE["SEND_APP_PAYLOAD_TO_STORAGE"] = "SEND_APP_PAYLOAD_TO_STORAGE";
    CRUD_APP_STAGE["SAVE_APP_TO_CACHE"] = "SAVE_APP_TO_CACHE";
    CRUD_APP_STAGE["DELETE_APP_FROM_CACHE"] = "DELETE_APP_FROM_CACHE";
    CRUD_APP_STAGE["GET_APP_DEPLOY_STATUS"] = "GET_APP_DEPLOY_STATUS";
    CRUD_APP_STAGE["DELETE_SUCCESSFUL"] = "DELETE_SUCCESSFUL";
    CRUD_APP_STAGE["CREATE_SUCCESSFUL"] = "CREATE_SUCCESSFUL";
    CRUD_APP_STAGE["UPDATE_SUCCESSFUL"] = "UPDATE_SUCCESSFUL";
})(CRUD_APP_STAGE || (exports.CRUD_APP_STAGE = CRUD_APP_STAGE = {}));
