// @tria-sdk/core/utils/constants.ts
import { ENV } from '../types';

export const SUPPORTED_EVM_CHAINS = [
  'ETH',
  'POLYGON',
  'AVALANCHE',
  'ARBITRUM',
  'BINANCE',
  'OPTIMISM',
  'FANTOM',
];

export const ALLCHAINS = [
  'ETH',
  'POLYGON',
  'AVALANCHE',
  'ARBITRUM',
  'BINANCE',
  'OPTIMISM',
  // 'KLAYTN',
  'FANTOM',
  // 'CELO',
  // 'GNOSIS',
  // 'OKX',
  // 'SOLANA',
  // 'APTOS',
  // 'SUI',
  // 'BITCOIN',
  // 'LITECOIN',
  // 'NEO',
  // 'POLKADOT',
  // 'STATEMINT',
  // 'MOONBEAM',
  // 'COSMOS',
  // 'SEI',
];

export const isEvmChain = (chainName: string) => {
  return SUPPORTED_EVM_CHAINS.includes(chainName);
};

export const isChainSupported = (chainName: string) => {
  return ALLCHAINS.includes(chainName);
};

const SUBGRAPH_URL = {
  testnet: 'https://api.thegraph.com/subgraphs/name/saurabh-threely/triadids',
  'mainnet-staging':
    'https://api.thegraph.com/subgraphs/name/saurabh-threely/indexer-stagingmainnet',
  mainnet: 'https://api.thegraph.com/subgraphs/name/saurabh-threely/indexer-mainnet',
};

export const getSubgraphUrl = (environment: ENV = 'mainnet') => {
  return SUBGRAPH_URL[environment];
};
