// @tria-sdk/web/dist/services/nonevm.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import BackendService from "./index";
import { errorHandler } from "../utils/helper";
class nonEVMService extends BackendService {
    constructor(apiUrl) {
        super(apiUrl);
    }
    send(shard1, chainName, payload) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const { data } = yield this.triaAxios.post("/nonevm/send", {
                    shard1: shard1,
                    chainName: chainName,
                    payload: payload
                });
                return data;
            }
            catch (err) {
                return errorHandler(err);
            }
        });
    }
    swap(shard1, chainName, payload) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const { data } = yield this.triaAxios.post("/nonevm/swap", {
                    shard1: shard1,
                    chainName: chainName,
                    payload: payload
                });
                return data;
            }
            catch (err) {
                return errorHandler(err);
            }
        });
    }
}
export default nonEVMService;
