// @tria-sdk/core/dist/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSubgraphUrl = exports.isChainSupported = exports.isEvmChain = exports.ALLCHAINS = exports.SUPPORTED_EVM_CHAINS = void 0;
exports.SUPPORTED_EVM_CHAINS = [
    'ETH',
    'POLYGON',
    'AVALANCHE',
    'ARBITRUM',
    'BINANCE',
    'OPTIMISM',
    'FANTOM',
];
exports.ALLCHAINS = [
    'ETH',
    'POLYGON',
    'AVALANCHE',
    'ARBITRUM',
    'BINANCE',
    'OPTIMISM',
    // 'KLAYTN',
    'FANTOM',
    // 'CELO',
    // 'GNOSIS',
    // 'OKX',
    // 'SOLANA',
    // 'APTOS',
    // 'SUI',
    // 'BITCOIN',
    // 'LITECOIN',
    // 'NEO',
    // 'POLKADOT',
    // 'STATEMINT',
    // 'MOONBEAM',
    // 'COSMOS',
    // 'SEI',
];
const isEvmChain = (chainName) => {
    return exports.SUPPORTED_EVM_CHAINS.includes(chainName);
};
exports.isEvmChain = isEvmChain;
const isChainSupported = (chainName) => {
    return exports.ALLCHAINS.includes(chainName);
};
exports.isChainSupported = isChainSupported;
const SUBGRAPH_URL = {
    testnet: 'https://api.thegraph.com/subgraphs/name/saurabh-threely/triadids',
    'mainnet-staging': 'https://api.thegraph.com/subgraphs/name/saurabh-threely/indexer-stagingmainnet',
    mainnet: 'https://api.thegraph.com/subgraphs/name/saurabh-threely/indexer-mainnet',
};
const getSubgraphUrl = (environment = 'mainnet') => {
    return SUBGRAPH_URL[environment];
};
exports.getSubgraphUrl = getSubgraphUrl;
