// @tria-sdk/web/dist/utils/helper/constants.js
export const covalentChainNames = {
    ETH: 'eth-mainnet',
    POLYGON: 'matic-mainnet',
    AVALANCHE: 'avalanche-mainnet',
    ARBITRUM: 'arbitrum-mainnet',
    BINANCE: 'bsc-mainnet',
    OPTIMISM: 'optimism-mainnet',
    // KLAYTN: 'klaytn-mainnet',
    FANTOM: 'fantom-mainnet',
    MOONBEAM: 'moonbeam-mainnet',
    SOLANA: 'solana-mainnet',
    BITCOIN: 'btc-mainnet',
};
// GNOSIS
// CELO
export const explorerBaseUrls = {
    ETH: 'https://etherscan.io',
    POLYGON: 'https://polygonscan.com',
    AVALANCHE: 'https://snowtrace.io',
    ARBITRUM: 'https://arbiscan.io',
    BINANCE: 'https://bscscan.com',
    OPTIMISM: 'https://optimistic.etherscan.io',
    FANTOM: 'https://ftmscan.com',
    SOLANA: 'https://solscan.io',
    APTOS: 'https://explorer.aptoslabs.com',
    NEO: 'https://neo3.neotube.io',
    SUI: 'https://suiexplorer.com/?network=mainnet',
    FUSE: 'https://explorer.fuse.io',
};
export const txnBaseUrl = {
    ETH: `${explorerBaseUrls.ETH}/tx`,
    POLYGON: `${explorerBaseUrls.POLYGON}/tx`,
    AVALANCHE: `${explorerBaseUrls.AVALANCHE}/tx`,
    ARBITRUM: `${explorerBaseUrls.ARBITRUM}/tx`,
    BINANCE: `${explorerBaseUrls.BINANCE}/tx`,
    OPTIMISM: `${explorerBaseUrls.OPTIMISM}/tx`,
    FANTOM: `${explorerBaseUrls.FANTOM}/tx`,
    FUSE: `${explorerBaseUrls.FUSE}/tx`,
};
export const ALLCHAINS = [
    'ETH',
    'POLYGON',
    'Avalanche',
    'ARBITRUM',
    'BINANCE',
    'OPTIMISM',
    'KLAYTN',
    'FANTOM',
    'CELO',
    'GNOSIS',
    'SOLANA',
    'APTOS',
    'SUI',
    'BITCOIN',
    'LITECOIN',
    'NEO',
    'POLKADOT',
    // 'STATEMINT',
    // 'MOONBEAM',
    // 'COSMOS',
    // 'SEI',
    'FUSE',
];
export const CHAINS_FOR_COVALENT = [
    'ETH',
    'POLYGON',
    'AVALANCHE',
    'ARBITRUM',
    'BINANCE',
    'OPTIMISM',
    // 'KLAYTN',
    'FANTOM',
    'MOONBEAM',
    'SOLANA',
    'BITCOIN',
];
export const CHAINS_FOR_CLIENT_SIDE = [
    'ETH',
    'POLYGON',
    'AVALANCHE',
    'ARBITRUM',
    'BINANCE',
    'OPTIMISM',
    'KLAYTN',
    'FANTOM',
    'CELO',
    'GNOSIS',
    'FUSE',
];
export const CHAINS_FOR_SERVER_SIDE = [
    'SOLANA',
    'APTOS',
    'SUI',
    'BITCOIN',
    'LITECOIN',
    'NEO',
    'POLKADOT',
    // 'STATEMINT',
    // 'MOONBEAM',
    // 'COSMOS',
];
export const nativeTokenDetails = {
    ETH: {
        name: 'Ether',
        symbol: 'ETH',
        logoUrl: 'https://www.datocms-assets.com/86369/1669619533-ethereum.png',
    },
    POLYGON: {
        name: 'Matic Token',
        symbol: 'MATIC',
        logoUrl: 'https://logos.covalenthq.com/tokens/1/0x7d1afa7b718fb893db30a3abc0cfc608aacfebb0.png',
    },
    AVALANCHE: {
        name: 'Avalanche Coin',
        symbol: 'AVAX',
        logoUrl: 'https://www.datocms-assets.com/86369/1686152997-avalanche-colour.png',
    },
    ARBITRUM: {
        name: 'Arbitrum Mainnet Ether',
        symbol: 'ETH',
        logoUrl: 'https://www.datocms-assets.com/86369/1669925028-arbitrum.png',
    },
    BINANCE: {
        name: 'Binance Coin',
        symbol: 'BNB',
        logoUrl: 'https://www.datocms-assets.com/86369/1670003837-biance-smart-chain-colour.png',
    },
    OPTIMISM: {
        name: 'Ether',
        symbol: 'ETH',
        logoUrl: 'https://www.datocms-assets.com/86369/1670347461-optimisim-colour.png',
    },
    FANTOM: {
        name: 'Fantom',
        symbol: 'FTM',
        logoUrl: 'https://www.datocms-assets.com/86369/1669925359-fantom-1.png',
    },
    FUSE: {
        name: 'Fuse',
        symbol: 'FUSE',
        logoUrl: 'https://assets.coingecko.com/coins/images/10347/standard/fuse.png',
    },
};
